src/dune_rules/lock_rules.ml

```diff
@@ -0,0 +1,125 @@
+open Import
+open Memo.O
+
+module Spec = struct
+  type ('path, 'target) t =
+    { target : 'target
+    ; lock_dir : string
+    }
+
+  let name = "lock"
+  let version = 1
+  let bimap t _ g = { t with target = g t.target }
+  let is_useful_to ~memoize = memoize
+
+  let encode { target; lock_dir } _encode_path encode_target : Sexp.t =
+    Sexp.List [ encode_target target; Sexp.Atom lock_dir ]
+  ;;
+
+  let action { target; lock_dir } ~ectx:_ ~eenv:_ =
+    let open Fiber.O in
+    let+ () = Fiber.return () in
+    Printf.eprintf
+      "Running ACTION, target is %s, our lock_dir is %S\n"
+      (Path.Build.to_string target)
+      lock_dir;
+    let path = Path.build target in
+    Path.mkdir_p path;
+    let t = Unix.localtime @@ Unix.gettimeofday () in
+    let content =
+      sprintf
+        "Created on %d-%0.2d-%0.2d %0.2d:%0.2d:%0.2d"
+        (t.tm_year + 1900)
+        (t.tm_mon + 1)
+        t.tm_mday
+        t.tm_hour
+        t.tm_min
+        t.tm_sec
+    in
+    Io.write_file ~binary:true (Path.relative path "lock.dune") content;
+    ()
+  ;;
+end
+
+module A = Action_ext.Make (Spec)
+
+let action ~target ~lock_dir = A.action { Spec.target; lock_dir }
+
+let lock ~target ~lock_dir =
+  action ~target ~lock_dir
+  |> Action.Full.make ~can_go_in_shared_cache:false
+  |> Action_builder.With_targets.return
+  |> Action_builder.With_targets.add_directories ~directory_targets:[ target ]
+;;
+
+module Gen_rules = Build_config.Gen_rules
+
+let rule ?loc { Action_builder.With_targets.build; targets } =
+  (* TODO this ignores the workspace file *)
+  Rule.make ~info:(Rule.Info.of_loc_opt loc) ~targets build |> Rules.Produce.rule
+;;
+
+let setup_lock_rules ctx_name ~lock_dir ~projects : Gen_rules.result =
+  let sctx = Super_context.find_exn ctx_name in
+  let target =
+    let ( / ) = Path.Build.relative in
+    Private_context.t.build_dir
+    / Context_name.to_string ctx_name
+    / ".lock"
+    / lock_dir
```

Comment:
Instead of composing the target from the component, you should make it relative to the `~dir` you were passed in `gen_rules`. This is the only valid directory you can use for targets.

Reply:
That's a good point, I've changed it and it is much nicer now.